In [ ]:
from google.colab import drive
drive.mount('/content/gdrive/')

Mounted at /content/gdrive/


In [ ]:
!ls /content/gdrive/

MyDrive


In [ ]:
import numpy as np
import pandas as pd
import scipy.io as io
from sklearn import metrics
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Lasso, Ridge, SGDRegressor, SGDClassifier


In [ ]:
dataX = pd.read_csv("/content/gdrive/MyDrive/ex1data2.txt", header=None, usecols=[0, 1], sep=',') #change the path

dataY = pd.read_csv("/content/gdrive/MyDrive/ex1data2.txt", header=None, usecols=[2], sep=',') #change the path
data_X = np.asarray(dataX)

data_Y = np.asarray(dataY)

In [ ]:
X_train, X_test, Y_train, Y_test = train_test_split(data_X, data_Y, test_size=0.3,random_state=1)

print(X_train)

[[1811    4]
 [2526    3]
 [1888    2]
 [4478    5]
 [4215    4]
 [1236    3]
 [2567    4]
 [3000    4]
 [2200    3]
 [1268    3]
 [1940    4]
 [2238    3]
 [1839    2]
 [1000    1]
 [1664    2]
 [1767    3]
 [2609    4]
 [1100    3]
 [1534    3]
 [1427    3]
 [ 852    2]
 [1600    3]
 [1320    2]
 [2104    3]
 [2300    4]
 [1985    4]
 [2000    3]
 [1494    3]
 [1380    3]
 [1890    3]
 [1200    3]
 [2132    4]]


In [ ]:
from sklearn import metrics

def calc_metrics(true, pred):
    mse = metrics.mean_squared_error(true, pred)
    rmse = np.sqrt(mse)
    mae = metrics.mean_absolute_error(true, pred)
    r2=metrics.r2_score(true, pred) #Коэффициент детерминации измеряет долю дисперсии.Фактически, данная мера качества — это нормированная среднеквадратичная ошибка. Если r2 близка к единице, то модель хорошо объясняет данные.
    
    print("RMSE:  {}\nMAE:   {} \nR2:   {}".format(rmse, mae, r2))

In [ ]:
model_linear = Lasso(alpha = 0.1)
model_linear.fit(X_train, Y_train)
print(model_linear.coef_, model_linear.intercept_)

[  139.68898838 -8757.05182985] [90421.82230494]


In [ ]:
y_pred_linear = model_linear.predict(X_test)

calc_metrics(Y_test, y_pred_linear)

(15, 1) (15,)
RMSE:  78874.29073397799
MAE:   58956.36624163459 
R2:   0.639480238808477


In [ ]:

X_train_new = np.hstack((Y_train**0, X_train))

X_test_new = np.hstack((Y_test**0, X_test))


A = X_train_new
Y = Y_train
At = np.transpose(A)
Aminus = At.dot(A)

n = np.dot(At,A).shape
number = n[0]
b = np.eye(number)

alpha = 0.1


answer = np.dot(np.dot(np.linalg.pinv(np.dot(At,A)),At), Y) #формула коэфф-ов лин регрессии (х(t)*х)^(-1)*x(t)*y
print ("Тета:",answer)

Y_pred = X_test_new.dot(answer) #считаем прогноз

print("Prognoz:", Y_pred)


calc_metrics(Y_test, Y_pred) #сравниваем прогноз с реальными значениями и, исходя из этого, считаем ошибки



Тета: [[90422.30665191]
 [  139.68918288]
 [-8757.32850439]]
Prognoz: [[607541.24253416]
 [288211.77047506]
 [357401.00601659]
 [264883.67693443]
 [399404.36004594]
 [270707.53259838]
 [432510.69638803]
 [340358.92570548]
 [314097.35932441]
 [267817.14977487]
 [478790.90593757]
 [237225.21872459]
 [232196.40814098]
 [502355.28782703]
 [329463.16944099]]
RMSE:  78874.29109731178
MAE:   58956.34710050829 
R2:   0.6394802354870142


LR: RMSE:  2.862794230037707
MAE:   2.282771866322345 
R2:   0.2648616239903123

Lasso: RMSE:  2.8498191109414024
MAE:   2.253021041936627 
R2:   0.27151029755777756

Ridge: RMSE:  2.862523334584119
MAE:   2.282233880797087 
R2:   0.2650007441684432

МНК с регуляризацией: RMSE:  78870.16957631338
MAE:   58932.27820944658 
R2:   0.6395179119196526

МНК:RMSE:  78874.29109731178 MAE:   58956.34710050829  R2:   0.6394802354870142

In [ ]:
from sklearn import preprocessing

X_train_new = np.hstack((Y_train**0, X_train))

X_test_new = np.hstack((Y_test**0, X_test))

#минимаксная нормализация
X_train_norm = preprocessing.normalize(X_train_new,norm = 'l2', axis = 0 )
X_test_norm = preprocessing.normalize(X_test_new,norm = 'l2', axis = 0 )

Y_test_norm = preprocessing.normalize(Y_test,norm = 'l2', axis = 0 )
Y_train_norm = preprocessing.normalize(Y_train,norm = 'l2', axis = 0 )

 

A = X_train_norm
Y = Y_train_norm
At = np.transpose(A)
Aminus = At.dot(A)


answer = np.dot(np.dot(np.linalg.pinv(np.dot(At,A)),At), Y) #формула коэфф-ов лин регрессии (х(t)*х)^(-1)*x(t)*y
print ("Тета:",answer)

Y_pred = X_test_norm.dot(answer) #считаем прогноз

print("Prognoz:", Y_pred)


calc_metrics(Y_test_norm, Y_pred) #сравниваем прогноз с реальными значениями и, исходя из этого, считаем ошибки



In [ ]:
#мнк для трёхмерной
X_train_new = np.hstack((Y_train**0, X_train))

X_test_new = np.hstack((Y_test**0, X_test))


A = X_train_new
Y = Y_train
At = np.transpose(A)
Aminus = At.dot(A)


answer = np.dot(np.dot(np.linalg.pinv(np.dot(At,A)),At), Y) #формула коэфф-ов лин регрессии (х(t)*х)^(-1)*x(t)*y
print ("Тета:",answer)

Y_pred = X_test_new.dot(answer) #считаем прогноз

print("Prognoz:", Y_pred)


calc_metrics(Y_test, Y_pred) #сравниваем прогноз с реальными значениями и, исходя из этого, считаем ошибки



In [ ]:
#мнк для трёхмерной c регуляризацией
X_train_new = np.hstack((Y_train**0, X_train))

X_test_new = np.hstack((Y_test**0, X_test))


A = X_train_new
Y = Y_train
At = np.transpose(A)
Aminus = At.dot(A)

n = np.dot(At,A).shape
number = n[0]
b = np.eye(number)

alpha = 0.1


answer = np.dot(np.dot(np.linalg.pinv(np.dot(At,A)+alpha*b),At), Y) #формула коэфф-ов лин регрессии (х(t)*х)^(-1)*x(t)*y
print ("Тета:",answer)

Y_pred = X_test_new.dot(answer) #считаем прогноз

print("Prognoz:", Y_pred)


calc_metrics(Y_test, Y_pred) #сравниваем прогноз с реальными значениями и, исходя из этого, считаем ошибки



In [ ]:
from sklearn import preprocessing

X_train_new = np.hstack((Y_train**0, X_train))

X_test_new = np.hstack((Y_test**0, X_test))

#стандартизация
scaler = StandardScaler().fit(X_train_new)
X_train_scaled = scaler.transform(X_train_new)

scaler_1 = StandardScaler().fit(X_test_new)
X_test_scaled = scaler_1.transform(X_test_new) 

scaler_2 = StandardScaler().fit(Y_test)
Y_test_scaled = scaler_2.transform(Y_test)

scaler_3 = StandardScaler().fit(Y_train)
Y_train_scaled = scaler_3.transform(Y_train)

A = X_train_scaled
Y = Y_train_scaled
At = np.transpose(A)
Aminus = At.dot(A)


answer = np.dot(np.dot(np.linalg.pinv(np.dot(At,A)),At), Y) #формула коэфф-ов лин регрессии (х(t)*х)^(-1)*x(t)*y
print ("Тета:",answer)

Y_pred = X_test_scaled.dot(answer) #считаем прогноз

print("Prognoz:", Y_pred)


calc_metrics(Y_test_scaled, Y_pred) #сравниваем прогноз с реальными значениями и, исходя из этого, считаем ошибки

